kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/EngineYarnModeSubmitter.scala

```diff
@@ -0,0 +1,451 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kyuubi.engine.deploy
+
+import java.io.{File, FileOutputStream, InterruptedIOException, IOException, OutputStreamWriter}
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.util
+import java.util.{Locale, Properties}
+import java.util.zip.{ZipEntry, ZipOutputStream}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.security.UserGroupInformation
+import org.apache.hadoop.yarn.api.ApplicationConstants
+import org.apache.hadoop.yarn.api.ApplicationConstants.Environment
+import org.apache.hadoop.yarn.api.records._
+import org.apache.hadoop.yarn.client.api.{YarnClient, YarnClientApplication}
+import org.apache.hadoop.yarn.exceptions.ApplicationNotFoundException
+import org.apache.hadoop.yarn.util.Records
+
+import org.apache.kyuubi.{KyuubiException, Logging, Utils}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf._
+import org.apache.kyuubi.engine.deploy.EngineYarnModeSubmitter._
+import org.apache.kyuubi.util.KyuubiHadoopUtils
+
+abstract class EngineYarnModeSubmitter extends Logging {
+
+  val KYUUBI_ENGINE_STAGING: String = ".kyuubiEngineStaging"
+
+  val LOCALIZED_LIB_DIR = "__kyuubi_engine_libs__"
+  val LOCALIZED_CONF_DIR = "__kyuubi_engine_conf__"
+
+  val KYUUBI_CONF_FILE = "kyuubi-defaults.conf"
```

Comment:
this part seems ported from spark, the original comments are too brief, I would like to see a more detailed comment to help developers understand the directory layout

Reply:
SGTM